protocol/javatests/org/openyolo/protocol/AdditionalPropertiesTest.java

```diff
@@ -0,0 +1,222 @@
+
+    /**
+     * Classloader shenanigans within Robolectric mean that we can't just directly instantiate
+     * and provide the instances of AdditionalPropertiesBuilder via the @Parameters annotated
+     * method. We must ensure that the instances are only created once the test methods are
+     * actually executing, as it is only at this point that the robolectric shadows have been
+     * loaded. So, as an awkward workaround, this map holds lambdas to create the builders
+     * on-demand, and the tests are parameterized by the keys of the map.
+     */
+    private static final Map<String, Callable<AdditionalPropertiesBuilder>> buildersByContainerType =
+            ImmutableMap.<String, Callable<AdditionalPropertiesBuilder>>builder()
+                    .put("Credential",
+                            () -> new Credential.Builder(
+                                    ValidFacebookCredential.ID,
+                                    ValidFacebookCredential.AUTHENTICATION_METHOD,
+                                    ValidFacebookCredential.AUTHENTICATION_DOMAIN))
+                    .put("CredentialDeleteRequest",
+                            () -> new CredentialDeleteRequest.Builder(
+                                    ValidFacebookCredential.make()))
+                    .put("CredentialDeleteResult",
+                            () -> new CredentialDeleteResult.Builder(
+                                    CredentialDeleteResult.CODE_DELETED))
+                    .put("CredentialRetrieveRequest",
+                            () -> new CredentialRetrieveRequest.Builder(
+                                    AuthenticationMethods.EMAIL))
+                    .put("CredentialRetrieveResult",
+                            () -> new CredentialRetrieveResult.Builder(
+                                    CredentialRetrieveResult.CODE_NO_CREDENTIALS_AVAILABLE))
+                    .put("CredentialSaveRequest",
+                            () -> new CredentialSaveRequest.Builder(
+                                    ValidFacebookCredential.make()))
+                    .put("CredentialSaveResult",
+                            () -> new CredentialSaveResult.Builder(
+                                    CredentialSaveResult.CODE_SAVED))
+                    .put("Hint",
+                            () -> new Hint.Builder(
+                                    ValidFacebookCredential.ID,
+                                    ValidFacebookCredential.AUTHENTICATION_METHOD))
+                    .put("HintRetrieveRequest",
+                            () -> new HintRetrieveRequest.Builder(
+                                    AuthenticationMethods.EMAIL))
+                    .put("HintRetrieveResult",
+                            () -> new HintRetrieveResult.Builder(
+                                    HintRetrieveResult.CODE_NO_HINTS_AVAILABLE))
+            .build();
+
+    @Parameters(name = "AdditionalPropertiesTest for {0}")
```

Comment:
I didn't know this existed, interesting feature.

Reply:
Yeah, it's new to me too actually. The way of declaring the test parameters is a little ugly and as noted in the comment involves lots of unexpected behavior, but it was worth it for this change.